lib/core/proxy.c

```diff
@@ -673,9 +713,19 @@ static h2o_httpclient_body_cb on_head(h2o_httpclient_t *client, const char *errs
     /* switch to using pipe reader, if the opportunity is provided */
     if (args->pipe_reader != NULL) {
 #ifdef __linux__
-        if (pipe2(self->pipe_reader.fds, O_NONBLOCK | O_CLOEXEC) != 0) {
-            char errbuf[256];
-            h2o_fatal("pipe2(2) failed:%s", h2o_strerror_r(errno, errbuf, sizeof(errbuf)));
+        struct pipe_list *pl = &req->conn->ctx->proxy.pipe_list;
+        if (!h2o_linklist_is_empty(&pl->anchor)) {
+            struct pipe_reader *pr = H2O_STRUCT_FROM_MEMBER(struct pipe_reader, link, pl->anchor.next);
+            h2o_linklist_unlink(&pr->link);
+            self->pipe_reader.fds[0] = pr->fds[0];
```

Comment:
I wonder if a check here that the pipe is empty makes sense (with `recv(MSG_PEEK)`). Maybe behind a `#if __has_feature(address_sanitizer)`?

Reply:
i'm not sure -- i think we could do that... IMO draining the pipe so that it is empty before adding it to the list should be enough, but i think if we wanted to add an additional check in certain cases that would be OK.

Reply:
Yes, i agree that draining the pipe should be enough for correctness. I'm thinking this as a debug assertion, but i don't feel strongly about it.

Reply:
I have a mixed feeling. Yes, we can add something like that.

At the same time, I think the question is if having such kind of mechanism helps us improve quality, when we have so many file descriptors, with some of them connected to clients and servers. Is there a reason to believe that we would have higher probability of writing to these pipes than to other file descriptors?